### PR TITLE
support constraints in plot contour

### PR DESCRIPTION
**Pull request by y-kamiya**
## Motivation
https://github.com/optuna/optuna/issues/4829
Make `plot_contour` plots support optimization with constraints

## Description of the changes
show feasible or infeasible plots by changing marker color
 
スクリーンショット 2023-09-30 15 04 53

sample code to show plots above
```
import optuna
from optuna.samplers._base import _CONSTRAINTS_KEY

def objective(trial):
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_categorical("y", [-1, 0, 1])
    trial.set_user_attr(_CONSTRAINTS_KEY, [y])
    return x ** 2 + y

def constraints(trial):
    return trial.user_attrs[_CONSTRAINTS_KEY]

sampler = optuna.samplers.TPESampler(seed=10, constraints_func=constraints)
study = optuna.create_study(sampler=sampler)
study.optimize(objective, n_trials=30)

fig = optuna.visualization.plot_contour(study, params=["x", "y"])
fig.show()
```

### CODE REVIEW

File: optuna/visualization/_contour.py [hunk 1]
```diff
@@ -226,16 +241,23 @@ def _get_contour_subplot(
...
+        _create_scatter(feasible.x, feasible.y, "black"),
+        _create_scatter(infeasible.x, infeasible.y, "#cccccc"),
+    )
+
+
+def _create_scatter(x: list[Any], y: list[Any], marker_color: str) -> Scatter:
+    return go.Scatter(
+        x=x,
+        y=y,
+        marker={"line": {"width": 2.0, "color": "Grey"}, "color": marker_color},
```
Comment (keisuke-umezawa) [inline optuna/visualization/_contour.py:255]: Here, `{"width": 2.0, "color": "Grey"}`, the line color of infeasible is also marker_color.

**Review by not522 — COMMENTED** [review]
Thank you for your PR! Could you check my comment and fix the lint error?

**Comment by keisuke-umezawa** [comment]
@y-kamiya 
If you fix the commented part and lint, it looks good to me! Could you change this PR accordingly?

**Comment by y-kamiya** [comment]
@keisuke-umezawa 
thanks to check PR. i will fix lint in this week.

File: optuna/visualization/_contour.py [hunk 2]
```diff
@@ -226,16 +241,28 @@ def _get_contour_subplot(
...
+    marker_color = "black" if is_feasible else "#cccccc"
+    return go.Scatter(
+        x=x,
+        y=y,
+        marker={
+            "line": {"width": 2.0, "color": edge_color},
+            "color": marker_color,
+        },
         mode="markers",
+        name="Infeasible Trial",
```
[the remark this answers is not in the document]
Comment (y-kamiya) [reply optuna/visualization/_contour.py:262]: do you mean `infeasible` is not correct word? 
if so, could you tell me what is correct definition of word of feasiblity in optuna?

[the remark this answers is not in the document]
Comment (not522) [reply optuna/visualization/_contour.py:262]: In the current implementation, all trials are named "Infeasible Trial". If they meet the constraints, please set them to "Feasible Trial".

[the remark this answers is not in the document]
Comment (y-kamiya) [reply optuna/visualization/_contour.py:262]: oh, you meant it's a bug, thank you!

**Review by not522 — APPROVED** [review]
Thank you for your update! LGTM!

**Review by keisuke-umezawa — APPROVED** [review]
Thank you for working on it. I merge it.

**Pull request merged**